src/unsync/mpsc.rs

```diff
@@ -0,0 +1,198 @@
+/// sending a message through this channel performs no dynamic allocation.
+pub fn channel<T>(buffer: usize) -> (Sender<T>, Receiver<T>) { channel_(Some(buffer)) }
+
+fn channel_<T>(buffer: Option<usize>) -> (Sender<T>, Receiver<T>) {
+    let shared = Rc::new(RefCell::new(Shared {
+        buffer: VecDeque::new(),
+        capacity: buffer,
+        blocked_senders: VecDeque::new(),
+        blocked_recv: None,
+    }));
+    let sender = Sender { shared: Rc::downgrade(&shared) };
+    let receiver = Receiver { shared: shared };
+    (sender, receiver)
+}
+
+struct Shared<T> {
+    buffer: VecDeque<T>,
+    capacity: Option<usize>,
+    blocked_senders: VecDeque<Task>,
+    blocked_recv: Option<Task>,
+}
+
+/// The transmission end of a channel.
+///
+/// This is created by the `channel` function.
+pub struct Sender<T> {
+    shared: Weak<RefCell<Shared<T>>>,
+}
+
+impl<T> Drop for Sender<T> {
+    fn drop(&mut self) {
+        if let Some(task) = self.shared.upgrade().and_then(|shared| shared.borrow_mut().blocked_recv.take()) {
+            // Wake up receiver to avoid deadlock in case this was the last sender
+            task.unpark();
+        }
+    }
+}
+
+impl<T> Clone for Sender<T> {
+    fn clone(&self) -> Self { Sender { shared: self.shared.clone() } }
+}
+
+impl<T> Sink for Sender<T> {
+    type SinkItem = T;
+    type SinkError = SendError<T>;
+
+    fn start_send(&mut self, msg: T) -> StartSend<T, SendError<T>> {
+        if let Some(shared) = self.shared.upgrade() {
+            let mut shared = shared.borrow_mut();
+
+            match shared.capacity {
+                Some(capacity) if shared.buffer.len() == capacity => {
+                    shared.blocked_senders.push_back(task::park());
+                    Ok(AsyncSink::NotReady(msg))
+                }
+                _ => {
+                    shared.buffer.push_back(msg);
+                    if let Some(task) = shared.blocked_recv.take() {
+                        task.unpark();
+                    }
+                    Ok(AsyncSink::Ready)
+                }
+            }
+        } else {
+            Err(SendError(msg))
+        }
+    }
+
+    fn poll_complete(&mut self) -> Poll<(), SendError<T>> {
+        Ok(Async::Ready(()))
+    }
+}
+
+
+/// The receiving end of a channel which implements the `Stream` trait.
+///
+/// This is created by the `channel` function.
+pub struct Receiver<T> {
+    shared: Rc<RefCell<Shared<T>>>,
+}
+
+impl<T> Drop for Receiver<T> {
+    fn drop(&mut self) {
+        let mut shared = self.shared.borrow_mut();
+        for task in shared.blocked_senders.drain(..) {
+            task.unpark();
+        }
+    }
+}
+
+impl<T> Stream for Receiver<T> {
+    type Item = T;
+    type Error = ();            // TODO: Replace with uninhabited type, as this can never fail
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        if let Some(shared) = Rc::get_mut(&mut self.shared) {
```

Comment:
I think this succeeding means all senders are gone, right? If so, could you add a comment to that effect?